src/main/scala/net/snowflake/spark/snowflake/SnowflakeRelation.scala

```diff
@@ -182,115 +183,118 @@ private[snowflake] case class SnowflakeRelation(
     resultSchema: StructType
   ): RDD[T] = {
     val conn = DefaultJDBCWrapper.getConnector(params)
-    Utils.genPrologueSql(params).foreach(x => x.execute(bindVariableEnabled = false)(conn))
-    Utils.executePreActions(DefaultJDBCWrapper, conn, params, params.table)
-    Utils.setLastSelect(statement.toString)
-    log.info(s"Now executing below command to read from snowflake:\n${statement.toString}")
-
-    val startTime = System.currentTimeMillis()
-    val (resultSet, queryID, serializables) = try {
-      if (params.isExecuteQueryWithSyncMode) {
-        val rs = statement.execute(bindVariableEnabled = false)(conn)
-        val queryID = rs.asInstanceOf[SnowflakeResultSet].getQueryID
-        log.info(s"The query ID for reading from snowflake is: $queryID; " +
-          s"The query ID URL is:\n${params.getQueryIDUrl(queryID)}")
-        val objects = rs
-          .asInstanceOf[SnowflakeResultSet]
-          .getResultSetSerializables(params.expectedPartitionSize)
-        (rs, queryID, objects)
-      } else {
-        val asyncRs = statement.executeAsync(bindVariableEnabled = false)(conn)
-        val queryID = asyncRs.asInstanceOf[SnowflakeResultSet].getQueryID
-        log.info(s"The query ID for async reading from snowflake is: $queryID; " +
-          s"The query ID URL is:\n${params.getQueryIDUrl(queryID)}")
-        SparkConnectorContext.addRunningQuery(sqlContext.sparkContext, conn, queryID)
-        // The query is executed in async mode, getResultSetSerializables() is blocked
-        // until query is done.
-        val objects = asyncRs
-          .asInstanceOf[SnowflakeResultSet]
-          .getResultSetSerializables(params.expectedPartitionSize)
-        SparkConnectorContext.removeRunningQuery(sqlContext.sparkContext, conn, queryID)
-        (asyncRs, queryID, objects)
+    try {
```

Comment:
The actual code change is to put the code segment with `try {...} finally {conn.close}` to make sure connection is closed if query failed.